source/common/event/scaled_timer.cc

```diff
@@ -0,0 +1,207 @@
+  return active_;
+}
+
+ScaledTimerManager::ScaledTimerManager(Dispatcher& dispatcher)
+    : dispatcher_(dispatcher),
+      dispatch_timer_(dispatcher.createTimer([this] { dispatchCallback(); })), timers_(1.0) {}
+
+Event::TimerPtr ScaledTimerManager::createTimer(TimerCb callback) {
+  return std::make_unique<ScaledTimerImpl>(*this, std::move(callback));
+}
+
+void ScaledTimerManager::setDurationScaleFactor(double scale_factor) {
+  timers_.setScaleFactor(scale_factor);
+  internalScheduleTimer();
+}
+
+ScaledTimerManager::ScaledTimerList::ScaledTimerList(double scale_factor)
+    : timers_(Compare{.scale_factor = scale_factor}) {}
+
+void ScaledTimerManager::ScaledTimerList::addActive(ScaledTimerImpl* timer) {
+  ASSERT(timer->active().has_value());
+  timers_.insert(timer);
+}
+
+void ScaledTimerManager::ScaledTimerList::removeActive(ScaledTimerImpl* timer) {
+  ASSERT(timer->active().has_value());
+  timers_.erase(timer);
+}
+
+void ScaledTimerManager::ScaledTimerList::setScaleFactor(double scale_factor) {
+  std::set<ScaledTimerImpl*, Compare> reordered(timers_.begin(), timers_.end(),
+                                                Compare{.scale_factor = scale_factor});
+  std::swap(reordered, timers_);
+}
+
+absl::optional<ScaledTimerManager::ScaledTimerImpl*>
+ScaledTimerManager::ScaledTimerList::pop(MonotonicTime now) {
+  if (timers_.empty()) {
+    return absl::nullopt;
+  }
+
+  auto* timer = *timers_.begin();
+  if (timers_.key_comp().targetTime(timer) <= now) {
+    timers_.erase(timers_.begin());
+    return timer;
+  }
+  return absl::nullopt;
+}
+
+absl::optional<MonotonicTime> ScaledTimerManager::ScaledTimerList::firstScheduleTime() const {
+  if (timers_.empty()) {
+    return absl::nullopt;
+  }
+  return timers_.key_comp().targetTime(*timers_.begin());
+}
+
+MonotonicTime
+ScaledTimerManager::ScaledTimerList::Compare::targetTime(const ScaledTimerImpl* timer) const {
+  ASSERT(timer->active().has_value());
+
+  auto scaled_delay = timer->active()->delay * scale_factor;
+  return timer->active()->start +
+         std::chrono::duration_cast<std::chrono::microseconds>(scaled_delay);
+}
+
+bool ScaledTimerManager::ScaledTimerList::Compare::operator()(const ScaledTimerImpl* lhs,
+                                                              const ScaledTimerImpl* rhs) const {
+  return targetTime(lhs) < targetTime(rhs);
+}
+
+void ScaledTimerManager::scheduleTimer(ScaledTimerImpl* timer) {
+  ASSERT(timer->active().has_value());
+
+  timers_.addActive(timer);
+  internalScheduleTimer();
+}
+
+void ScaledTimerManager::descheduleTimer(ScaledTimerImpl* timer) {
+  ASSERT(timer->active().has_value());
+  timers_.removeActive(timer);
+  internalScheduleTimer();
+}
+
+void ScaledTimerManager::dispatchCallback() {
+  // Find all timers that are ready to be handled and dispatch them.
+  const auto now = dispatcher_.timeSource().monotonicTime();
+  for (auto timer = timers_.pop(now); timer.has_value(); timer = timers_.pop(now)) {
```

Comment:
Consider this instead:
do {
  auto timer = timers_.pop(now);
  if (!timer.has_value()) {
    break;
  }
  timer.value()->active_.reset();
  timer.value()->dispatch(dispatcher_);
} while(true);